### PR TITLE
CUDA: mmq CLI option, fixed mmq build issues

### PR DESCRIPTION
**Pull request by JohannesGaessler**
This PR aims to fix the following:

* The new quantized matrix matrix multiplication kernels decide whether or not to unroll a loop based on compute capability but with cmake this check is not functional by default. This PR adds PTX code for compute capability 7.0.
* For GPUs with compute capability < 6.1 the new quantized matrix matrix multiplication kernels cannot be used due to a lack of the `__dp4a` intrinsic. This PR adds a corresponding check.
* For the `perplexity` binary cuBLAS should always be used as suggested by https://github.com/ggerganov/llama.cpp/pull/2160#issuecomment-1657079913 . I tried to add a corresponding check to cmake but adding the define at that position does nothing. Is it actually possible to do this with cmake?

### CODE REVIEW

**Comment by Green-Sky** [comment]
> Is it actually possible to do this with cmake?

this is going to be hard with single config cmake projects, if not impossible.
for ci, you could run cmake + build 2 times, one without ppl and one only ppl.

**Comment by slaren** [comment]
About the cuBLAS issue, I think it may be better to make it a command line option, there isn't much reason for this to be a compile-time for now, since we still have to link to cuBLAS regardless

**Comment by Green-Sky** [comment]
> since we still have to link to cuBLAS regardless

we do?

**Comment by JohannesGaessler** [comment]
Currently yes because the KV cache uses f16 for which there is no mul_mat_q implementation.

**Comment by JohannesGaessler** [comment]
Okay, so the reason I made this a compile option instead of a command line option is because I did not want to add temporary changes to the command line options. How about this: a command line option like `--mul-mat-algorithm` that takes values like `blas`, `mul_mat_q` (open for better names), or `auto` (decide based on current performance level). If at some point we remove cuBLAS as a mandatory dependency and the user compiles without cuBLAS the behavior of `auto` is adjusted and explicitly requesting `cuBLAS` results in an error.

**Comment by slaren** [comment]
We already have 2 temporary parameters for llama 2, so I don't think adding another one would be too bad. My opinion is that something like `--mul-mat-algorithm` shouldn't be a parameter in the long term either, this is not something that the user should have to decide. So if it is going to be a temporary parameter regardless, it may as well be a simple cublas yes/cublas no.

**Comment by JohannesGaessler** [comment]
I changed the new kernels to be opt-in. They can be used with the CLI option `--mul-mat-q`.

File: README.md [hunk 1]
```diff
@@ -402,7 +402,9 @@ Building the program with BLAS support may lead to some performance improvements
 
   | Option                  | Legal values           | Default | Description |
   |-------------------------|------------------------|---------|-------------|
+<!---
   | LLAMA_CUDA_CUBLAS       | Boolean                |   false | Use cuBLAS instead of custom CUDA kernels for prompt processing. Faster for all quantization formats except for q4_0 and q8_0, especially for k-quants. Increases VRAM usage (700 MiB for 7b, 970 MiB for 13b, 1430 MiB for 33b). |
+--->
```
Comment (slaren) [inline README.md:407]: This breaks the table formatting. You can preview the rendered file here by clicking on "view file".

Comment (JohannesGaessler) [reply README.md]: Thank you. I forgot to check.

**Review by slaren — APPROVED** [review]

**Comment by slaren** [comment]
Quick test with 3090 Ti  under WSL2:

| Model | cuBLAS | mmq |
| -- | -- | -- |
| llama-2-7b.ggmlv3.q4_0.bin |  1473.09 | 1544.40 |
| llama-2-7b.ggmlv3.q5_K_M.bin | 1461.38 | 875.14 |

**Comment by LostRuins** [comment]
I think line https://github.com/ggerganov/llama.cpp/blob/0728c5a8b9569183ffca0399caac099afef87595/CMakeLists.txt#L269 should be `GGML_CUDA_F16` as `GGML_CUDA_DMMV_F16` does not seem to exist in the code.

**Comment by LostRuins** [comment]
Also, performance wise I did observe a consistent speedup for q4_0, and a slowdown for q4_k_m (the 2 formats that I usually use)

**Comment by JohannesGaessler** [comment]
>I don't know how you did it in such a short time, but you did it.

I did it by doing nothing else for an entire day because all of my physics colleagues are on vacation.

>Also, performance wise I did observe a consistent speedup for q4_0, and a slowdown for q4_k_m (the 2 formats that I usually use)

Which GPU do you have?

**Comment by Dampfinchen** [comment]
I've made a similar observation. While the speed for the regular quants is great, the slowdown for k-quants is definately very noticeable. Here is a 7b model with 20 GPU layers, comparing q4_0 and q4k_m on an RTX 2060 laptop:

7b4_0 20 layers
7bq4k_m

**Comment by LostRuins** [comment]
> Which GPU do you have?

I have a laptop Nvidia RTX 2060 6GB. 

7B q4_K_M, Prompt processing on 1968 tokens, 32 layers offloaded. 

non-mmq takes 14.7s (7ms/T)
mmq takes 18.5s (9ms/T)

**Comment by JohannesGaessler** [comment]
Can you try disabling [this loop unroll](https://github.com/ggerganov/llama.cpp/blob/0728c5a8b9569183ffca0399caac099afef87595/ggml-cuda.cu#L2876-L2878)? I only have an RTX 3090 and several Pascal cards on which the unroll was significantly slower (by a factor of 4). So it would be useful for me if someone with an RTX 2000 GPU could actually test the unroll (but so far it looks like it's not an issue).

**Comment by LostRuins** [comment]
Disabling the loop unroll seemed to have a very slight negative performance impact:

**Enabled: 19.1s (9ms/T)
Disabled: 20.1s (10ms/T)**

However there was another interesting side effect: the binary size changed significantly, removing the highlighted loop unrolling code:
image
reduced the output binary file size by nearly 15%! 

Nonetheless, both approaches are still significantly inferior to old method (non-MMQ):

**Non-MMQ: 15.3s (8ms/T)**

more concerning is the mmq changes breaking compatibility with full-offload offload for non-standard (vocab!=32000) models.

**Comment by JohannesGaessler** [comment]
>more concerning is the mmq changes breaking compatibility with full-offload offload for non-standard (vocab!=32000) models, even when mmq is disabled

I know what the problem is, I'll fix it soon.

**Comment by Dampfinchen** [comment]
@JohannesGaessler Alright, so I can confirm without a shadow of a doubt, the new mul-mat is faster than cublas at processing regular quants, but slower for k-quants. Here is the data to back it up:

First, let's start with 7b, 20 layers offloaded on an RTX2060 and q4_0.

This is with MMQ:

4_0mmq

This is without MMQ:

4_0nommq

Notice prompt processing speed is a little faster with the new mul mat.

Now, let's switch over to q4k_m.

First, MMQ enabled:

q4k_m_mulmatq

This is with regular cublas:

q4k_m-nommq

Now you will notice with k-quants its a different story. Prompt processing is slower than cuBLAS.

Regarding loop unroll, I've been similar results to LostRuins.

If you need me to test anything, feel free to ask! I'm happy to help.

**Pull request merged**